### PR TITLE
docs: refactor TLS Context docs to make it more reader friendly

### PR DESCRIPTION
**Pull request by agrawroh**
## Description

I noticed that some of the sections of the TLS Context docs are very verbose and hard to read. This PR is to make it a little bit more user-friendly.

Screenshot 2024-12-10 at 11 54 52

---

**Commit Message:** docs: refactor TLS Context docs to make it more reader friendly
**Additional Description:** Refining the TLS context docs to make it more user-friendly.
**Risk Level:** N/A
**Testing:** N/A
**Docs Changes:** N/A
**Release Notes:** N/A

### CODE REVIEW

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 1]
```diff
@@ -70,16 +68,19 @@ message UpstreamTlsContext {
...
 
-  // This field is used to control the enforcement, whereby the handshake will fail if the keyUsage extension
-  // is present and incompatible with the TLS usage. Currently, the default value is false (i.e., enforcement off)
-  // but it is expected to be changed to true by default in a future release.
-  // ``ssl.was_key_usage_invalid`` in :ref:`listener metrics <config_listener_stats>` will be set for certificate
-  // configurations that would fail if this option were set to true.
+  // Controls enforcement of the keyUsage extension in peer certificates. If set to true, the handshake will fail if
+  // the keyUsage is incompatible with TLS usage.
+  //
+  // **Note:** The default value is ``false`` (i.e., enforcement off). It is expected to change to ``true`` in a future
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:79]:
```suggestion
  // .. note::
  //   The default value is ``false`` (i.e., enforcement off). It is expected to change to ``true`` in a future
```

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 2]
```diff
@@ -70,16 +68,19 @@ message UpstreamTlsContext {
...
   //
   // Defaults to 1, setting this to 0 disables session resumption.
   google.protobuf.UInt32Value max_session_keys = 4;
 
-  // This field is used to control the enforcement, whereby the handshake will fail if the keyUsage extension
-  // is present and incompatible with the TLS usage. Currently, the default value is false (i.e., enforcement off)
-  // but it is expected to be changed to true by default in a future release.
-  // ``ssl.was_key_usage_invalid`` in :ref:`listener metrics <config_listener_stats>` will be set for certificate
-  // configurations that would fail if this option were set to true.
+  // Controls enforcement of the keyUsage extension in peer certificates. If set to true, the handshake will fail if
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:76]:
```suggestion
  // Controls enforcement of the ``keyUsage`` extension in peer certificates. If set to ``true``, the handshake will fail if
```

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 3]
```diff
@@ -70,16 +68,19 @@ message UpstreamTlsContext {
...
+  // the keyUsage is incompatible with TLS usage.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:77]:
```suggestion
  // the ``keyUsage`` is incompatible with TLS usage.
```

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 4]
```diff
@@ -139,46 +132,49 @@ message DownstreamTlsContext {
     bool disable_stateless_session_resumption = 7;
   }
 
-  // If set to true, the TLS server will not maintain a session cache of TLS sessions. (This is
-  // relevant only for TLSv1.2 and earlier.)
+  // If true, the TLS server will not maintain a session cache of TLS sessions.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:135]:
```suggestion
  // If ``true``, the TLS server will not maintain a session cache of TLS sessions.
```

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 5]
```diff
@@ -139,46 +132,49 @@ message DownstreamTlsContext {
...
+  //
+  // **Note:** This applies only to TLSv1.2 and earlier.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:137]:
```suggestion
  // .. note::
  //   This applies only to TLSv1.2 and earlier.
```

Comment (phlax) [reply api/envoy/extensions/transport_sockets/tls/v3/tls.proto:137]: same elsewhere

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 6]
```diff
@@ -139,46 +132,49 @@ message DownstreamTlsContext {
...
+  // Allows multiple certificates in the Downstream transport socket for different SNI.
+  // If the client provides an SNI with no matching certificate, Envoy will decide to perform a full certificate scan
+  // based on this config.
+  // Defaults to ``false``. See more details in :ref:`Multiple TLS certificates <arch_overview_ssl_cert_select>`.
   google.protobuf.BoolValue full_scan_certs_on_sni_mismatch = 9;
 
-  // By default, Envoy as a server uses its preferred cipher during the handshake.
-  // Setting this to true would allow the downstream client's preferred cipher to be used instead.
-  // Has no effect when using TLSv1_3.
+  // If true, allows the downstream client's preferred cipher to be used during the handshake. If false, Envoy defaults
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:161]:
```suggestion
  // If ``true``, allows the downstream client's preferred cipher to be used during the handshake. If ``false``, Envoy defaults
```

**Review by phlax — COMMENTED** [review]
thanks for working on this @agrawroh - some format/markup nits ...

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 7]
```diff
@@ -70,16 +68,20 @@ message UpstreamTlsContext {
...
-  // is present and incompatible with the TLS usage. Currently, the default value is false (i.e., enforcement off)
-  // but it is expected to be changed to true by default in a future release.
-  // ``ssl.was_key_usage_invalid`` in :ref:`listener metrics <config_listener_stats>` will be set for certificate
-  // configurations that would fail if this option were set to true.
+  // Controls enforcement of the ``keyUsage`` extension in peer certificates. If set to ``true``, the handshake will fail if
+  // the ``keyUsage`` is incompatible with TLS usage.
+  //
+  // .. note::
+  //   The default value is ``false`` (i.e., enforcement off). It is expected to change to ``true`` in a future
+  // release.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:81]: @agrawroh i think its this - `release.` needs to be either indented or on the previous line

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 8]
```diff
@@ -139,46 +133,50 @@ message DownstreamTlsContext {
     bool disable_stateless_session_resumption = 7;
   }
 
-  // If set to true, the TLS server will not maintain a session cache of TLS sessions. (This is
-  // relevant only for TLSv1.2 and earlier.)
+  // If ``true``, the TLS server will not maintain a session cache of TLS sessions.
+  //
+  // .. note::
+  //   This applies only to TLSv1.2 and earlier.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:139]: and/or this - needs a blank line after

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 9]
```diff
@@ -139,46 +133,50 @@ message DownstreamTlsContext {
...
   google.protobuf.BoolValue full_scan_certs_on_sni_mismatch = 9;
 
-  // By default, Envoy as a server uses its preferred cipher during the handshake.
-  // Setting this to true would allow the downstream client's preferred cipher to be used instead.
-  // Has no effect when using TLSv1_3.
+  // If ``true``, allows the downstream client's preferred cipher to be used during the handshake. If ``false``, Envoy defaults
+  // to using its preferred cipher.
+  //
+  // .. note::
+  //   This has no effect when using TLSv1_3.
```
Comment (phlax) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:166]: and here

**Comment by phlax** [comment]
needs api signoff

cc @ggreenway for TLS docs update

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 10]
```diff
@@ -51,14 +50,13 @@ message UpstreamTlsContext {
   // interacts with other validation options.
   bool auto_host_sni = 6;
 
-  // If true, replace any Subject Alternative Name validations with a validation for a DNS SAN matching
-  // the SNI value sent. Note that the validation will be against the actual requested SNI, regardless of how it
-  // is configured.
+  // If true, replaces any Subject Alternative Name (SAN) validations with a validation for a DNS SAN matching
+  // the SNI value sent. The validation uses the actual requested SNI, regardless of its configuration.
```
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:54]: I think the final sentence rewording here makes it less clear. There are 3-4 different ways the final SNI can be computed via different configuration and runtime, so `its` here doesn't have a clear antecedent.

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 11]
```diff
@@ -51,14 +50,13 @@ message UpstreamTlsContext {
...
   //
-  // For the common case where an SNI value is sent and it is expected that the server certificate contains a SAN
-  // matching that SNI value, this option will do the correct SAN validation.
+  // This option ensures correct SAN validation when an SNI value is provided and expected to match the server
```
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:56]: I think it should somehow note that this is the common/normal case

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 12]
```diff
@@ -139,46 +132,53 @@ message DownstreamTlsContext {
...
+  //
+  // .. note::
+  //   This applies only to TLSv1.2 and earlier.
+  //
   bool disable_stateful_session_resumption = 10;
 
-  // If specified, ``session_timeout`` will change the maximum lifetime (in seconds) of the TLS session.
-  // Currently this value is used as a hint for the `TLS session ticket lifetime (for TLSv1.2) <https://tools.ietf.org/html/rfc5077#section-5.6>`_.
-  // Only seconds can be specified (fractional seconds are ignored).
+  // Maximum lifetime of TLS sessions in seconds. If specified, ``session_timeout`` will change the maximum lifetime
```
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:142]:
```suggestion
  // Maximum lifetime of TLS sessions. If specified, ``session_timeout`` will change the maximum lifetime
```

Duration has units, so no need for `in seconds`

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 13]
```diff
@@ -139,46 +132,53 @@ message DownstreamTlsContext {
...
+  // Defaults to ``LENIENT_STAPLING``
   OcspStaplePolicy ocsp_staple_policy = 8 [(validate.rules).enum = {defined_only: true}];
 
-  // Multiple certificates are allowed in Downstream transport socket to serve different SNI.
-  // If the client provides SNI but no such cert matched, it will decide to full scan certificates or not based on this config.
-  // Defaults to false. See more details in :ref:`Multiple TLS certificates <arch_overview_ssl_cert_select>`.
+  // Allows multiple certificates in the Downstream transport socket for different SNI.
+  // If the client provides an SNI with no matching certificate, Envoy will decide to perform a full certificate scan
+  // based on this config.
+  // Defaults to ``false``. See more details in :ref:`Multiple TLS certificates <arch_overview_ssl_cert_select>`.
```
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:160]:
```suggestion
  // Multiple certificates are allowed in Downstream transport socket to serve different SNI.
  // This option controls the behavior when no matching certificate is found for the received SNI value,
  // or no SNI value was sent. If enabled, all certificates will be evaluated for a match for non-SNI criteria
  // such as key type and OCSP settings. If disabled, the first provided certificate will be used.
  // Defaults to ``false``. See more details in :ref:`Multiple TLS certificates <arch_overview_ssl_cert_select>`.
```

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto [hunk 14]
```diff
@@ -139,46 +132,53 @@ message DownstreamTlsContext {
...
   google.protobuf.BoolValue full_scan_certs_on_sni_mismatch = 9;
 
-  // By default, Envoy as a server uses its preferred cipher during the handshake.
-  // Setting this to true would allow the downstream client's preferred cipher to be used instead.
-  // Has no effect when using TLSv1_3.
+  // If ``true``, allows the downstream client's preferred cipher to be used during the handshake. If ``false``, Envoy defaults
+  // to using its preferred cipher.
+  //
+  // .. note::
+  //   This has no effect when using TLSv1_3.
```
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/tls.proto:167]:
```suggestion
  // If ``true``, the downstream client's preferred cipher is used during the handshake. If ``false``, Envoy
  // uses its preferred cipher.
  //
  // .. note::
  //   This has no effect when using TLSv1_3.
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
Thanks for cleaning up these docs!

/wait

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto (hunk 10 above)
Comment (agrawroh) [reply api/envoy/extensions/transport_sockets/tls/v3/tls.proto:54]: Will change it to say, _"The validation uses the actual requested SNI, regardless of how the SNI is configured."_

File: api/envoy/extensions/transport_sockets/tls/v3/tls.proto (hunk 11 above)
Comment (agrawroh) [reply api/envoy/extensions/transport_sockets/tls/v3/tls.proto:56]: Changed it to, _"For common cases where an SNI value is present and the server certificate should include a corresponding SAN, this option ensures the SAN is properly validated."_

**Pull request merged**